rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+  </tr>
+</table>
+
+<table>
+  <tr>
+   <td><strong>Input[3]:</strong>
+<p>
+<code><em># Next, SchemaGen without a run / read.</em></code>
+<code>infer_schema = SchemaGen(stats=statistics_gen.outputs['output'])</code>
+<p>
+<code><em># Finally, ExampleValidator with a run / read. Note that SchemaGen</em></code>
+<code><em># will be implicitly run (see Note 2 below).</em></code>
+<code>validate_stats = ExampleValidator(</code>
+<p>
+<code>    stats=statistics_gen.outputs['output'],</code>
+<p>
+<code>    schema=infer_schema.outputs['output'])</code>
+<p>
+<code>context.run(validate_stats)</code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[3]:</strong>
+<p>
+_ExecutionResult object for ExampleValidator_
+   </td>
+  </tr>
+</table>
+
+Note that the user may have forgotten to run InteractiveContext.run() on
+upstream components in the dependency graph. Instead of implicitly running these
+upstream components, we remind the user to explicitly run upstream notebook
+cells (with a readable error message). We think this explicit mode of component
+execution is more notebook-friendly, and is easy to use with common notebook
+actions such as “Run All”, “Run Cells Before”, and “Run Cells After”.
+
+### Export to a selected orchestration engine (v0)
+
+#### Filter out InteractiveContext() objects
+
+##### Option 1: Replace InteractiveContext instances with dummy versions.
+
+1.  Search for possible import alias, e.g. `from
+    tfx.orchestration.interactive.interactive_context import InteractiveContext
+    as FooBar`
+
+*   Search for all instances of string ".*InteractiveContext", or the alias name
+    if found from prior step.
+*   Replace each instance with `DummyInteractiveContext`, which inherits from
+    InteractiveContext and basically does nothing / returns empty
+    ExecutionResult on .run().
+
+    ```
+    class DummyInteractiveContext(InteractiveContext):
+    def run(self,
+      component: base_component.BaseComponent,
+      enable_cache: bool = True):
+    return None
+    ```
+
+    1.  This should cover the case where the class definition is aliased.
+
+        ```
+        aliased_class = interactive_context.InteractiveContext
+        sess = aliased_class()
+        ```
+
+    1.  This should cover subsequent aliases of InteractiveContext instances.
+
+        ```
+        a = InteractiveContext()
+        b = a
+        c = InteractiveContext()
+        d = c
+        ```
+
+Cons:
+
+*   DummyInteractiveContext is now present/clutters the production pipeline code
+    (it's a no-op so mainly affects readability, not execution).
+*   Down the line, converting back to a notebook (replacing
+    DummyInteractiveContext with InteractiveContext) could be fragile.
+
+##### Option 2: Ensure InteractiveContext only runs in notebook context. **[recommended]**
+
+*   If InteractiveContext is run outside of a notebook context, just log a
+    warning and return.
+*   Bi-directional import to notebook from pipeline would "just work".
```

Comment:
How is this an export at all? Doesn't this just mean the user has to rewrite the pipeline?